rmw_fastrtps_shared_cpp/src/rmw_get_topic_info.cpp

```diff
@@ -0,0 +1,306 @@
+  }
+  return ret;
+}
+
+
+rmw_ret_t
+_get_info_by_topic(
+  const char * identifier,
+  const rmw_node_t * node,
+  rcutils_allocator_t * allocator,
+  const char * topic_name,
+  bool no_mangle,
+  bool is_publisher,
+  rmw_topic_info_array_t * participants_info)
+{
+  rmw_ret_t ret = _validate_params(
+    identifier,
+    node,
+    allocator,
+    topic_name,
+    participants_info);
+  if (ret != RMW_RET_OK) {
+    return ret;
+  }
+
+  const auto & topic_fqdns = _get_topic_fqdns(topic_name, no_mangle);
+  auto impl = static_cast<CustomParticipantInfo *>(node->data);
+  // The GUID of the participant associated with this node
+  const auto & participant_guid = impl->participant->getGuid();
+  const auto & node_name = node->name;
+  const auto & node_namespace = node->namespace_;
+  ::ParticipantListener * slave_target = impl->listener;
+  auto & topic_cache =
+    is_publisher ? slave_target->writer_topic_cache : slave_target->reader_topic_cache;
+  {
+    std::lock_guard<std::mutex> guard(topic_cache.getMutex());
+    auto & topic_name_to_data = topic_cache().getTopicNameToTopicData();
+    std::vector<rmw_topic_info_t> topic_info_vector;
+    for (const auto & topic_name : topic_fqdns) {
+      const auto & it = topic_name_to_data.find(topic_name);
+      if (it != topic_name_to_data.end()) {
+        for (const auto & data : it->second) {
+          rmw_topic_info_t topic_info = rmw_get_zero_initialized_topic_info();
+          rmw_ret_t ret = _set_rmw_topic_info(
+            &topic_info,
+            participant_guid,
+            node_name,
+            node_namespace,
+            data,
+            slave_target,
+            allocator);
+          if (ret != RMW_RET_OK) {
+            // Free topic_info
+            _handle_topic_info_fini(&topic_info, allocator);
+            // Free all the space allocated to the previous topic_infos
+            for (auto & tinfo : topic_info_vector) {
+              _handle_topic_info_fini(&tinfo, allocator);
+            }
+            return ret;
+          }
+          // add rmw_topic_info_t to a vector
+          topic_info_vector.push_back(topic_info);
+        }
+      }
+    }
+
+    // add all the elements from the vector to rmw_topic_info_array_t
+    auto count = topic_info_vector.size();
+    ret = rmw_topic_info_array_init_with_size(participants_info, count, allocator);
+    if (ret != RMW_RET_OK) {
+      rmw_error_string_t error_message = rmw_get_error_string();
+      rmw_reset_error();
+      RMW_SET_ERROR_MSG_WITH_FORMAT_STRING(
+        "rmw_topic_info_array_init_with_size failed to allocate memory: %s", error_message.str);
+      // Free all the space allocated to the previous topic_infos
+      for (auto & tinfo : topic_info_vector) {
+        _handle_topic_info_fini(&tinfo, allocator);
+      }
+      return ret;
+    }
+    for (auto i = 0u; i < count; i++) {
+      participants_info->info_array[i] = topic_info_vector.at(i);
```

Comment:
@jaisontj do we need an `std:;vector` if we'll eventually just copy it into a primitive array?

Reply:
It doesn't look trivial to determine what size of a primitive array would need be allocated.

Reply:
That's because of how the loop is written. If iterators that refer to matching topic data were collected first, then it'd be straightforward. This somewhat connects to @ivanpauno 's previous comment [here](https://github.com/ros2/rmw_fastrtps/pull/336/files#r349145606). I can understand why it was done this way initially but the overhead is not necessary. I'm fine with deferring though.

Reply:
I'm also ok with deferring, but I would really like to see it cleaned up at some point. My experience has been that clean up things like this tend to get overlooked though.